Add new meeting toc options sorted by reference number and only ad hoc agendaitems

njohner opened the pull request:
We add two new toc types for meeting periods:
* only agendaitems with proposals, ordered by dossier reference number
* only ad hoc agendaitems, ordered by titles

I'm not quite sure whether this is what was wanted here? The proposed solution is only partially satisfactory as it relies on a second entry in the TOC json `toc_adhoc`, which can be used in the sablon template to have a different listing type (without displaying the reference number) for adhoc agendaitem elements. I modified the template accordingly, see below for example TOCs:

**Only agendaitems with proposals, ordered by reference number**
<img width="950" alt="screen shot 2019-01-09 at 13 36 22" src="https://user-images.githubusercontent.com/7374243/50899868-ab287400-1413-11e9-867b-028a57a1a2c7.png">

**Only ad hoc agendaitems, ordered by title**
<img width="949" alt="screen shot 2019-01-09 at 13 36 34" src="https://user-images.githubusercontent.com/7374243/50899866-ab287400-1413-11e9-98ac-b4e983ea5057.png">

Another solution could be to have two different templates. Or maybe only one new TOC type, with two listings, one for agendaitems with proposal and one for ad hoc?

resolves #5176

deiferni commented on opengever/meeting/toc/reference_number.py:
```diff
@@ -0,0 +1,48 @@
+import re
+
+
+def alphanum_key(string):
```
i'd like to have unit tests for just these functions. no plone at all should be necessary, just a plain `UnitTest`.

deiferni commented on opengever/meeting/sablontemplate.py:
```diff
@@ -12,7 +12,7 @@
 VALIDATION_DATA = {
     'meeting': SAMPLE_MEETING_DATA,
-    'toc': {'toc': []},
+    'toc': {'toc': [], 'toc_adhoc': []},
```
as you already mention this is not satisfying as we would need to modify the templates. can't you just prepend the items to `toc`?

deiferni commented on opengever/meeting/toc/ad_hoc.py:
```diff
@@ -0,0 +1,47 @@
+from opengever.meeting.toc.alphabetical import AlphabeticalToc
+
+
+def first_title_char(item):
```
as with the other toc a `UnitTest` would be extra nice 😄.

deiferni commented on opengever/meeting/browser/toc.py:
```diff
@@ -124,3 +125,23 @@ def get_filename(self):
                       context=getRequest()))
+
+
+class DownloadAdHocAgendaitemTOC(DownloadAlphabeticalTOC):
```
i'm sorry the spec in the issue was garbled. the ad-hoc part should apply to both new tocs. please see https://github.com/4teamwork/opengever.core/issues/5176, i've updated it. there should be the following two tocs, 

- Gruppierung nach Positionsnummer der Ursprungs-Ordnungsposition
- Gruppierung nach Aktenzeichen des Ursprungs-Dossiers

in both of them you should filter out the ad-hocs and display them first, without a reference-number, as none exists.

njohner commented:
Ok I think I've gone a bit overboard here 😓. The code for `AlphabeticalToc`, which was used as baseclass for the `RepositoryBasedTOC` was a bit messy, with some unnecessary sorting operations (I think they were leftovers from before it was realised that the items needed to be sorted before the `groupby` statement) and the organization of the code made it not too easy to use as baseclass for other TOC classes. I therefore:
* refactored the `AlphabeticalToc` class a bit, moved `item_sort_key`and `group_by_key` to methods on the class, so that they can be easily overwritten. This allows to really just overwrite the sorting and grouping keys instead of the whole sorting and grouping methods.
* Moved some of the helper functions into a new `utlis.py` file
* adapted `RepositoryBasedTOC` to the new `AlphabeticalToc` baseclass 
* Added string normalisation before sorting to avoid things like having two different groups for titles starting with `A` or with `Ä`, sorting `Ä` after `Z`, sorting `A` after `z` and so on.
* Added the two new TOC classes (also based on `AlphabeticalToc`)
* Improved testing by adding `UnitTest`s for the helper functions 
* Added a group title for `Ad hoc proposals` in the `RepositoryBasedTOC` and the two new TOCs

**For example before in the `AlphabeticalToc` we had two different groups for `A` and `Ä`, now we have only one, and sorted as expected:**
<img width="949" alt="alphabetical_toc" src="https://user-images.githubusercontent.com/7374243/51462203-587a8080-1d60-11e9-9019-43dabacdac2d.png">

**In the `RepositoryBasedTOC` we would sort first all Upper case, then lower case, and then special characters:**
<img width="948" alt="repo_toc_old" src="https://user-images.githubusercontent.com/7374243/51462301-88c21f00-1d60-11e9-9670-9c95b81fcd18.png">
**Now we instead disregard case in sorting and normalise the special characters**
<img width="953" alt="repo_toc_new" src="https://user-images.githubusercontent.com/7374243/51462305-8bbd0f80-1d60-11e9-9d8d-f5a115596d05.png">

The new TOCs, there is one grouped by dossier reference number and one by repository number:
**grouped by repository number, sorting inside the group by title**
<img width="950" alt="screen shot 2019-01-21 at 09 42 35" src="https://user-images.githubusercontent.com/7374243/51462495-0f76fc00-1d61-11e9-8e5e-ecfee5aacf8d.png">
**grouped by dossier reference number**
<img width="962" alt="screen shot 2019-01-21 at 09 42 45" src="https://user-images.githubusercontent.com/7374243/51462502-1271ec80-1d61-11e9-836e-9bd03d3cebca.png">

deiferni commented on opengever/meeting/tests/test_toc.py:
```diff
@@ -10,12 +10,57 @@
+        string = u'A tEst'
+        self.assertEqual(normalise_string(string), u'a test')
+
+    def test_removes_umlauts(self):
```
not sure if we should do this. how does a swiss phonebook sort, or the libary 🤔? what about the "french" special chars.

deiferni replied:
as discusssed sort chars with diacritics like `ä` after their "base" char like `a`.

deiferni commented on opengever/meeting/toc/utils.py:
```diff
@@ -15,3 +16,30 @@ def normalise_string(string):
+
+def repo_refnum(item):
+    if item['dossier_reference_number'] is not None:
+        return item['dossier_reference_number'].split("/")[0].strip()
```
instead of hardcoding the `"/"` you should ask the current `IReferenceNumberFormatter` for its `repository_dossier_seperator`. theoretically it could be different between formatters.

njohner replied:
Oh did not realize that...

deiferni replied:
Good point, yes if it can be reused we should attempt to do so. IIRC it is tied closely to the plone content objects which we do not have access to at that moment. We can discuss it, if you want!

njohner commented:
Ok all done, though not 100% happy yet.
* I'm now using the sorter from `IReferenceNumberFormatter` to sort toc entries
  * This fixed sorting in the toc when using the `GroupedByThreeReferenceFormatter`, as there, simply casting to integers for the repository reference number does not give correct sorting.
  * Note that the sorter for `GroupedByThreeReferenceFormatter` ignores the `client_id`. This is not ideal in my opinion, as meetings can span several `client_id`s?
  * For the `RepositoryBasedTOC`, it seems that the repository titles are stored translated on the brains, so that if two agendaitems are generated from different languages, they will end up in different groups

I think the code for the `IReferenceNumberFormatter` would benefit from some refactoring. There seems to be a lot of duplicated code.

I think we could make an issue for a future release to refactor the `IReferenceNumberFormatter`s.

Apart from that, I think this is ready to merge.
